### ISSUE TITLE
How to turn off inline completion or prediction feature on ibus input method ?

### ISSUE THREAD

**Issue by Mahmudul-hasan613** [ibus/ibus#2420]
Which distribution and version?: 
Ubuntu 22.04 LTS

Which desktop environment and version?: 
ubuntu:GNOME

Which session type?: 
wayland

Which application and version?: 
All of software

IBus version?: 
IBus 1.5.26

Issue description:
When I typing with ibus, something like an underline that appears below the letter. Is there any option to stop it?

Steps to reproduce:
Check the screenshot

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
yes

Do you see any errors when you run ibus-daemon with the verbose option?:
No

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
yes

![Screenshot from 2022-06-28 00-22-07](https://user-images.githubusercontent.com/60663471/176009684-6c9a57b5-5569-4e5a-a17f-520c84288ac8.png)

**Comment by Mahmudul-hasan613** [comment]
I am using ibus-m17n. When I login to Gnome Xorg, how do I turn off the underline option?

**Comment by mike-fabian** [comment]
Screenshot of the ibus-m17n setup tool:

![Screenshot](https://user-images.githubusercontent.com/2330175/176614581-e6e85a7f-0ee4-48d0-bf0b-58acaa2e8965.png)

There are options for the underlining, one of the options is “None”.

**Comment by mike-fabian** [comment]
When you select “Hindi - itrans (m17n)” (or whatever ibus-m17n  input method you are using) in the Gnome panel, and then open the input menu of the Gnome panel again, you should see an entry “Setup” in that menu. Clicking on that should open the setup tool for “Hindi - itrans (m17n)”.

**Comment by mike-fabian** [comment]
When I look at the screenshot in your first comment, I have the impression that you are not using ibus-m17n but maybe ibus-typing-booster.

Because I see a rather long underlined text. ibus-m17n usually commits quite frequently, so you rarely seem more than one or two characters underlined. ibus-typing-booster commits only at the end of words so you may see the full word underline unless you switch the underlining off.

If you see a rocket icon in the gnome panel, you are using ibus-typing-booster.

If that is the case, you should also see a “Setup” entry in that menu in the Gnome panel. 

In the ibus-typing-booster setup tool, in the “Appearances” Tab, there is an option “Preedit underline” with a combobox with 4 options, one of them is for switching off the underline.

**Comment by mike-fabian** [comment]
The title of your issue also mentions “inline completion” and “prediction”. These are features only ibus-typing-booster has, ibus-m17n doesn’t have these features. This also makes me suspect that you are actually using ibus-typing-booster.

If that is the case, then look at this chapter in the ibus-typing-booster online documentation:

Simulate the behaviour of ibus-m17n: https://mike-fabian.github.io/ibus-typing-booster/docs/user/#2_2_1_1

It explains how to switch all the extra features off and how to switch the underline off.

This chapter also has a video which shows this in detail.

**Comment by mike-fabian** [comment]
Does it work for you now?

**Comment by Mahmudul-hasan613** [comment]
I use [OpenBangla](https://github.com/OpenBangla) keyboard software for typing Bangla. Which is based on ibus. But does it use ibus-typing-booster or ibus-m17n; I could not be sure. First I [discuss](https://github.com/OpenBangla/OpenBangla-Keyboard/discussions/296) the underline feature with OpenBangla developers. They first said it was an ibus bug, and I wanted to report it to ibus. But later I was told that the feature I wanted has not to the OpenBangla keyboard yet. However, this option will be considered later.
And the answer is : This is not work!

**Comment by mike-fabian** [comment]
No, OpenBangla keyboard does not use ibus-typing-booster or ibus-m17n. 
But you could try to use ibus-typying-booster to type Bangla as well.

**Comment by mike-fabian** [comment]
With ibus-typing-booster, you could use these 5 input methods for Bangla at the moment:

```
[mfabian@fedora ~]$ ls /usr/share/m17n/bn*
/usr/share/m17n/bn-disha.mim      /usr/share/m17n/bn-itrans.mim
/usr/share/m17n/bn-inscript.mim   /usr/share/m17n/bn-probhat.mim
/usr/share/m17n/bn-inscript2.mim
[mfabian@fedora ~]$ 
```

Maybe one of them is good for you?

**Comment by Mahmudul-hasan613** [comment]
Two things are very important to me.
1. I am used to typing in National (Jatiya) keyboard layout.
2. I need the Automatic Vowel Forming feature to type faster.

Do ibus-typing-booster have these features?

**Comment by mike-fabian** [comment]
> Two things are very important to me.
> 
>     1. I am used to typing in National (Jatiya) keyboard layout.
> 
>     2. I need the Automatic Vowel Forming feature to type faster.
> 
> 
> Do ibus-typing-booster have these features?

Probably not yet. I think none of the above 5 input methods currently available in m17n are “National (Jatiya)”. 

It would be possible to port this to m17n though. If anyone is interested to help me doing that, that would be nice.

**Comment by mike-fabian** [comment]
I just tested https://github.com/OpenBangla/OpenBangla-Keyboard on Fedora 36 Gnome Xorg.

It does *not* show an underline there:

![Screenshot-OpenBangla-Gnome-Xorg](https://user-images.githubusercontent.com/2330175/177145446-87cc789e-ced9-47b0-8fa4-130d6850c347.png)

**Comment by mike-fabian** [comment]
If I do exactly the same on Gnome Wayland, there is an underline:'


![Screenshot-OpenBangla-Gnome-Wayland](https://user-images.githubusercontent.com/2330175/177145535-f13affc0-6d82-48d0-b6cf-7dac9f46d821.png)

**Comment by mike-fabian** [comment]
I.e. ibus-openbangla does already try to switch off the underlining, only that doesn’t work on Wayland.

**Comment by mike-fabian** [comment]
> 2. I need the Automatic Vowel Forming feature to type faster.

Can you describe in detail what that feature does? Is there a specification somewhere?

**Comment by Mahmudul-hasan613** [comment]
> I just tested https://github.com/OpenBangla/OpenBangla-Keyboard on Fedora 36 Gnome Xorg.
> 
> It does _not_ show an underline there:

I logged in to Gnome Xorg according to your suggestion. But then the underline was showing while writing. I'm guessing, I may not have installed the operating system properly; Or an error has occurred in the system due to constant updates. I will completely reinstall the operating system and try again according to your suggestions; Then maybe work.

Thanks for your advice.

**Comment by Mahmudul-hasan613** [comment]
> Can you describe in detail what that feature does? Is there a specification somewhere?

Since I have no knowledge of coding; So I can't tell you anything technical about that feature. But I can give you an idea of ​​what this feature does.

The letters of Bengali and Hindi are generally divided into two parts. Consonants and vowels. There is another abbreviated form of the vowel. Which is called Dependent vowel signs in English. Which is very similar to Arabic Tashkil from ISO 8859-6. I have attached the PDF file of the Bengali Unicode list and the PDF file of the Arabic Unicode list.

[U0600 (Arabic Unicode list).pdf](https://github.com/ibus/ibus/files/9048482/U0600.Arabic.Unicode.list.pdf)
[U0980 (Bangla Unicode list).pdf](https://github.com/ibus/ibus/files/9048483/U0980.Bangla.Unicode.list.pdf)

If you look at these two files well, you can understand a lot. So the Automatic Vowel Forming feature works, where the vowel will be and Dependent vowel signs will be decided by the keyboard software itself.

[Mehdi Hasan Khan](https://mehdikhan.dev/) has worked with this feature. If you contact him, I hope you know the details.

He has created open source Bangla keyboard software for Windows operating system. Its name is [Avro keyboard](https://www.omicronlab.com/). The GitHub link is: https://github.com/omicronlab. The Linux version is the [Open Bangla keyboard](https://openbangla.github.io/).

This feature is also available in the [Gboard](https://play.google.com/store/apps/details?id=com.google.android.inputmethod.latin&gl=US) app of Android operating system.

**Comment by mike-fabian** [comment]
> > I just tested https://github.com/OpenBangla/OpenBangla-Keyboard on Fedora 36 Gnome Xorg.
> > It does _not_ show an underline there:
> 
> I logged in to Gnome Xorg according to your suggestion. But then the underline was showing while writing. I'm guessing, I may not have installed the operating system properly; Or an error has occurred in the system due to constant updates. I will completely reinstall the operating system and try again according to your suggestions; Then maybe work.

I don’t think completely reinstalling will help for whether the underline shows in Gnome Xorg or not, that might be just a waste of time.

**Comment by mike-fabian** [comment]
Could it be that you are using an older version of OpenBangla? When I tried it in Gnome Xorg and did not see the underline, I was using OpenBangla from current git master.

**Comment by Mahmudul-hasan613** [comment]
>  I don’t think completely reinstalling will help for whether the underline shows in Gnome Xorg or not, that might be just a waste of time.

I totally agree with you. But I can't find the problem! Besides the underline, there is another problem when I using Open Bangla keyboard. That is, the Ctrl + Backspace keyboard shortcut does not work when I am typing. After typing a word, I have to finish the text by pressing space button or arrow key or end button. Otherwise, when I move the cursor by mouse from the place where I was writing last, if I move the cursor elsewhere; Then that word also disappears, then wherever I put the mouse cursor, that word goes.

**Comment by mike-fabian** [comment]
I have created a  “National Jatiya” keyboard layout now which can be used with ibus-typing-booster or ibus-m17n, see:

https://github.com/mike-fabian/m17n-db-bn-national-jatiya

The README.md has instructions on how to install it.

This is a simple keyboard layout, just like `/usr/share/X11/xkb/symbol/bd`. It does *not* do any automatic vowel forming. Maybe I can add such an improvement later.

**Comment by Mahmudul-hasan613** [comment]
> Preedit disappears and appears elsewhere when the cursor is moved with a mouse. 
> https://github.com/OpenBangla/OpenBangla-Keyboard/issues/298

Many thanks for working on that issue.

**Comment by Mahmudul-hasan613** [comment]
> > 2. I need the Automatic Vowel Forming feature to type faster.
> 
> Can you describe in detail what that feature does? Is there a specification somewhere?

To get an idea about Automatic Vowel Forming, you need to understand Bengali or Hindi consonants, Independent vowels, Dependent vowel signs or Vowel Diacritics very well. Which you can find out from the GitHub link ==> https://github.com/lifeparticle/Bengali-Alphabet#phoneme

**Comment by Mahmudul-hasan613** [comment]
> Could it be that you are using an older version of OpenBangla? When I tried it in Gnome Xorg and did not see the underline, I was using OpenBangla from current git master.

No, I uninstalled today. And used that command to reinstall. == > bash -c "$(wget -q https://raw.githubusercontent.com/OpenBangla/OpenBangla-Keyboard/master/tools/install.sh -O -)"

When I tried it in Gnome Xorg and I see the underline.

**Comment by mike-fabian** [comment]
> > With ibus-typing-booster, you could use these 5 input methods for Bangla at the moment:
> > ```
> > [mfabian@fedora ~]$ ls /usr/share/m17n/bn*
> > /usr/share/m17n/bn-disha.mim      /usr/share/m17n/bn-itrans.mim
> > /usr/share/m17n/bn-inscript.mim   /usr/share/m17n/bn-probhat.mim
> > /usr/share/m17n/bn-inscript2.mim
> > [mfabian@fedora ~]$ 
> > ```
> >   
> > Maybe one of them is good for you?
> 
> I can't find anything!

You also need the m17n-db package:

https://packages.debian.org/bullseye/m17n-db

```
sudo apt-get install m17n-db
```

https://packages.debian.org/bullseye/all/m17n-db/filelist shows that the above files are in the m17n-db package.

**Comment by mike-fabian** [comment]
> Also shows the underline while typing even after the preedit underline is off. [Screencast from 7-7-22 12:25:53 পূর্বাহ্ণ +06.webm](https://user-images.githubusercontent.com/60663471/177618478-7bf566db-293b-48f6-8516-73c6abc9f28a.webm)

I see a red wiggly line under the first word “Bangla” in your screenshot. This makes me think you are using LibreOffice. The red wiggly line seems to be the spellchecking feature of LibreOffice.

Unfortunately, LibreOffice has the same problem as Wayland, it *always* shows the underline under the preedit, it cannot be disabled there, just like in Wayland. This is a bug in LibreOffice.

Here is a video where I type into LibreOffice and then into gedit using ibus-typing-booster with the underlining switched off. In the video you can see that LibreOffice ignores the setting and still shows the underline but in gedit it works:


https://user-images.githubusercontent.com/2330175/177718906-9489d672-1f8e-4418-a54a-e9539852acce.mp4

**Comment by Mahmudul-hasan613** [comment]
>  Here is a video where I type into LibreOffice and then into gedit using ibus-typing-booster with the underlining switched off. In the video you can see that LibreOffice ignores the setting and still shows the underline but in gedit it works:

I agree with you too. Shows the underline in LibreOffice, but not in the text editor. I checked myself after you said that.

**Comment by mike-fabian** [comment]
It is very weird that you have the `m17n-db` package installed but the files in `/usr/share/m17n` are not there.
I wonder how that is possible.

What happens if you try to list the files in the package with this command?:

```
dpkg -L m17n-db
```

**Comment by mike-fabian** [comment]
> And how to turn off dictionary or word suggestions on ibus-typing-booster?

See the chapter “Simulate the behaviour of ibus-m17n” in the ibus-typing-booster online documentation:

https://mike-fabian.github.io/ibus-typing-booster/docs/user/#2_2_1_1

1) Dictionaries and input methods tab:
   1a) Remove all input methods except the one you want to use
2) Options tab:
   2a) Check the option “Enable suggestions by key”
   2b) Check the option “Off the record mode”
3) Key bindings tab:
   3a) Remove all keys bound to the command “enable_lookup”
4) Appearance tab:
   4a) Set “Preedit underline” to “None”

If you skip 2b) and 3), then you can still get suggestions if you request them by typing `Tab` but you don’t get them by default anymore. If you do also 2b) and 3), you will never get any suggestions.

**Comment by Mahmudul-hasan613** [comment]
> It is very weird that you have the `m17n-db` package installed but the files in `/usr/share/m17n` are not there. I wonder how that is possible.
> 
> What happens if you try to list the files in the package with this command?:
> 
> ```
> dpkg -L m17n-db
> ```

![Screenshot from 2022-07-08 10-06-08](https://user-images.githubusercontent.com/60663471/177915748-c4048d50-4b6e-4141-83bf-3824716e65f2.png)

Writing that command brings up a long list. You can see that list in the video.
[Screencast from 8-7-22 10:06:59 পূর্বাহ্ণ +06.webm](https://user-images.githubusercontent.com/60663471/177915774-59a2580a-d8cf-41b6-87c5-c97297727c3f.webm)

**Comment by mike-fabian** [comment]
It is extremly weird that although the files are in the m17n-db package and `dpkg -L m17n-db` lists them correctly, they are not there if you check with `ls`. I was thinking whether there could be a typo in the command line using `ls`, but when I look at your screenshots of the setup tool of ibus-typing-booster, I see 

```
bn-inscript2   ❌ minput_open_im() failed 
```

This also means that the file `bn-inscript2.mim` is not available.

So the files really seem to be missing.

Maybe you get the files if you reinstall the `m17n-db` package?:

```
sudo apt-get install --reinstall m17n-db
```

?

**Comment by Mahmudul-hasan613** [comment]
The command you mentioned worked. But I can't add national (Jatiya) keyboard layout!

> bn-inscript2   ❌ minput_open_im() failed
> This also means that the file bn-inscript2.mim is not available.


Besides, the file is missing. How do I install it?
![Screenshot from 2022-07-08 23-31-02](https://user-images.githubusercontent.com/60663471/178042905-1255c0f0-8cf6-4267-a963-f5a7fee73ac8.png)
![Screenshot from 2022-07-08 23-31-13](https://user-images.githubusercontent.com/60663471/178042931-12375841-7c72-4894-9aad-6d8dbf7d53d3.png)

**Comment by mike-fabian** [comment]
I just made the national jatiya layout myself a few days ago.

It  is here:

https://github.com/mike-fabian/m17n-db-bn-national-jatiya

The README.md has instructions on how to install:

https://github.com/mike-fabian/m17n-db-bn-national-jatiya/blob/main/README.md

```
$ mkdir -p ~/.m17n.d/icons/
$ cp bn-national-jatiya.mim ~/.m17n.d/
$ cp bn-national-jatiya.png ~/.m17n.d/icons/
$ ibus restart
```

If you are using a Gnome session, `ibus restart` does not work, you have to log out of the Gnome session and log in again.

Instead of copying the `bn-national-jatiya.mim` to `~/.m17n.d/bn-national-jatiya.mim` and the icon to `~/.m17n.d/`bn-national-jatiya.png` you can also copy them to `/usr/share/m17n/bn-national-jatiya.mim` and `/usr/share/m17n/icons/bn-national-jatiya.png`, that achieves the same.

**Comment by mike-fabian** [comment]
I found a workaround to make switching off the underline in libreoffice off in a Gnome Xorg session.

First, the `libreoffice-kf5`  needs to be installed:

```
sudo apt-get install libreoffice-kf5
```

Then set this environment variable before starting libreoffice:

```
export SAL_USE_VCL_PLUGIN=kf5
```

and then start libreoffice:

```
oowriter &
```

Now switching off the underline works as you can see in this video:

(The “Low” underline makes a dotted underline, the “Double” underline looks the same as the “Single” underline, but “None” does indeed show no underline at all with `export SAL_USE_VCL_PLUGIN=kf5`)

https://user-images.githubusercontent.com/2330175/178066160-4d308d0b-f6d0-41cd-ba9f-2b6876c1830f.mp4

**Comment by Mahmudul-hasan613** [comment]
> I just made the national jatiya layout myself a few days ago.

G + C or AltGr + C = ঐ . But when I try it by ibus-typing-booster, its not work! I mean, I can't write Independent vowels through ibus-typing-booster's National (jatiya) keyboard layout.
![Screenshot from 2022-07-09 17-37-36](https://user-images.githubusercontent.com/60663471/178104282-699bce15-98a6-4346-a108-6de0c47d1f90.png)

**Comment by Mahmudul-hasan613** [comment]
I think the way the /usr/share/m17n/bn-unijoy.mim file is written; If you create the bn-national-jatiaya.mim file that way. Then it will work properly.

[bn-unijoy.txt](https://github.com/ibus/ibus/files/9076444/bn-unijoy.txt)

**Comment by mike-fabian** [comment]
> 

No, that has nothing to do with the problem you are seeing. Apparently the keyboard layout you  are using *before* switching to ibus-typing-booster has something on AltGr-c and AltGr-C  and when you press these keys, ibus-typing-booster does not get a AltGr+c, AltGr+C but some other stuff which bn-national-jatiaya.mim doesn’t contain and just passes through.

What keyboard layout are you using? Can you show this in a screenshot of your Gnome settings?

**Comment by mike-fabian** [comment]
bn-national-jatiya.mim contains:

```
((G-c) "এ") ; U+098F BENGALI LETTER E
((G-C) "ঐ") ; U+0990 BENGALI LETTER AI
```

i.e. if a `G-C` is coming as input, it is transliterated to   ঐ.

But if pressing AltGr+C  produces something else, this cannot work.

For example, if the keyboard layout “English (US, intl., with dead keys)” is used, pressing AltGr+C produces a 

¢ U+00A2 CENT SIGN

and bn-national-jatiya.mim does not transliterate ¢ to anything.

**Comment by mike-fabian** [comment]
Here is an example of 3 keyboard layouts which are slight variations of the English (US) layout and what they produce when typing AltGr+c, AltGr+C, and Alt+4 when used on their own (“NoIME” in the table) or when used with bn-national-jatiya.mim and ibus-typing-booster (“jatiya” in the table):

```
Keyboard layout                    |     AltGr+c     |     AltGr+C     |     AltGr+4     |
                                   | NoIME  | jatiya | NoIME  | jatiya | NoIME  | jatiya |
------------------------------------------------------------------------------------------
English (US)¹                      |   c    |   এ    |   C    |   ঐ   |   4     |   ৳    | 
English (India, with rupee)        |   c    |   এ    |   C    |   ঐ   |   ₹     |   ₹    | 
English (US, intl., with dead keys)|   ©    |   ©    |   ¢    |   ¢   |   ¤     |   ¤    |
------------------------------------------------------------------------------------------
¹ When English (US) is used, make sure to set Alternate Characters Key
  to “Right Alt” in the Gnome Control Center!!!
```

Screenshot showing how to set the Right Alt key as the AltGr key in the Gnome Control Center, this is important when using the English (US) keyboard layout!:

![Screenshot](https://user-images.githubusercontent.com/2330175/178110342-95721dac-ff1b-4507-950d-1ea66dbb5bd1.png)

**Comment by mike-fabian** [comment]
In the table in my last comment you see that when the “English (US,  intl., with dead keys)” layout is used, the AltGr+c, AltGr+C, and AltGr+4 keys don’t work correctly with bn-national-jatiya.mim. The reason is that this layout doesn’t produce G-c, G-C, and G-4 as input for bn-national-jatiya.mim but ©,  ¢, and ¤ respectively. And bn-national-jatiya.mim cannot transliterate this input and thus just passes it through unchanged. 

A similar effect happens with the “English (India, with rupee)” layout on the AltGr+4 key, a rupee symbol is produced there, i.e. the input to bn-national-jatiya.mim becomes ₹ which is not transliterated by bn-national-jatiya.mim and just passed through. bn-national-jatiya.mim  needs G-4 as the input to produce the desired ৳ U+09F3 BENGALI RUPEE SIGN result.

I suspect you are using a keyboard layout which produces something other than G-c and G-C when you type AltGr+c and AltGr+C.

As you see in the table, everything works fine with the standard “English (US)” layout so I recommend to use that one.

**Comment by mike-fabian** [comment]
A video showing how I type AltGr+c with and without ibus-typing-booster using first the “English (US)” keyboard layout and then the “English (US, intl., with dead keys)” layout:

https://user-images.githubusercontent.com/2330175/178110914-cd6cf369-b46f-402c-b9a6-dcad4731316c.mp4

**Comment by Mahmudul-hasan613** [comment]
After changing the gnome settings, it is now working properly.

**Comment by Mahmudul-hasan613** [comment]
> I found a workaround to make switching off the underline in libreoffice off in a Gnome Xorg session.
> 
> First, the `libreoffice-kf5` needs to be installed:
> 
> ```
> sudo apt-get install libreoffice-kf5
> ```
> 
> Then set this environment variable before starting libreoffice:
> 
> ```
> export SAL_USE_VCL_PLUGIN=kf5
> ```
> 
> and then start libreoffice:
> 
> ```
> oowriter &
> ```
> 
> Now switching off the underline works as you can see in this video:
> 
> (The “Low” underline makes a dotted underline, the “Double” underline looks the same as the “Single” underline, but “None” does indeed show no underline at all with `export SAL_USE_VCL_PLUGIN=kf5`)
> Peek.2022-07-08.22-30.mp4

I have followed the procedure you have shown. Then I gave None in place of Preedic underline option. But it still shows the underline when writing in LibreOffice.

**Comment by mike-fabian** [comment]
> After changing the gnome settings, it is now working properly.

Great!

About the automatic vowel forming: This is not yet done in bn-national-jatiya.mim, I may add something like that later ..., need to understand it better first!

**Comment by mike-fabian** [comment]
> > Then set this environment variable before starting libreoffice:
> > ```
> > export SAL_USE_VCL_PLUGIN=kf5
> >   
> > and then start libreoffice:
> > ```
> > oowriter &
> > ```
> I have followed the procedure you have shown. Then I gave None in place of Preedic underline option. But it still shows the underline when writing in LibreOffice.

Did you start libreoffice in the same terminal where you did do `export SAL_USE_VCL_PLUGIN=kf5`?

If you used a different terminal or started libreoffice using something like Alt+F2 or some menu, then SAL_USE_VCL_PLUGIN=kf5 might not be in the environment.

To make sure you can do this on a single command line like this:

```
env SAL_USE_VCL_PLUGIN=kf5 oowriter &
```

If done like that, it is sure that oowriter has SAL_USE_VCL_PLUGIN=kf5 in its enviroment.

If that works, you could add

```
export SAL_USE_VCL_PLUGIN=kf5
```

to your ~/.bashrc file, log out and  log in again, they you have it everywhere.

**Comment by Mahmudul-hasan613** [comment]
"oowriter &" command doesn't work.

**Comment by mike-fabian** [comment]
On Fedora, oowriter is just a startscript for libreoffice-writer:

```
$ rpm -qf /usr/bin/oowriter
libreoffice-writer-7.3.4.2-1.fc36.x86_64
$ file /usr/bin/oowriter
/usr/bin/oowriter: a /usr/bin/sh script, ASCII text executable
$ cat /usr/bin/oowriter 
#!/usr/bin/sh
exec libreoffice --writer "$@"
```
If that doesn’t exist on Ubuntu, you can use

```
env SAL_USE_VCL_PLUGIN=kf5 libreoffice --writer &

```

**Comment by Mahmudul-hasan613** [comment]
The Ubuntu command is "lowriter &". But even after opening in this way, the underline shows while typing.

**Comment by mike-fabian** [comment]
> The Ubuntu command is "lowriter &". But even after opening in this way, the underline shows while typing.

OK, then I have no idea why this workaround doesn’t work for you.

It is only a workaround anyway that disabling the underlining seems to work in the `kf5` backend for libreoffice but not in the `gtk3` backend.

We should probably open a bug against libreoffice and ask for proper support of the underlining style in the preedit. in both the `kf5` (where it works partly) and especially the `gtk3` backend (where it doesn’t work at all.

**Comment by mike-fabian** [comment]
I reported a bug against libreoffice here: https://bugs.documentfoundation.org/show_bug.cgi?id=149932

**Comment by Mahmudul-hasan613** [comment]
Thank you so much for working on that issue. It was very important to me. If the underline shows without the need while typing; Then it becomes a cause of annoyance.

**Comment by Mahmudul-hasan613** [comment]
> In the meantime, if you have any more feedback on bn-national-jatiya.mim and ibus-typing-booster, please tell me!

/ usr / share / m17n / bn * Bengali keyboard layouts with bn-unijoy.mim, gnome settings do not need to be changed to use AltGr key when typing with them. Also, when g + C is pressed, ঐ letter is supposed to appear. Which I can do in Open Bangla keyboard and bn-unijoy.mim layout. But that is not possible in the bn-national-jatiya.mim layout you created.

![Screenshot from 2022-07-11 16-46-35](https://user-images.githubusercontent.com/60663471/178249515-f21668db-d32e-4f53-b0e1-02fd0ae5970a.png)

![Screenshot from 2022-07-11 16-46-52](https://user-images.githubusercontent.com/60663471/178249544-052c0391-388e-47c4-9e71-b8f85770f82c.png)

**Comment by Mahmudul-hasan613** [comment]
Another thing I noticed. When I try to launch any software by searching in Ubuntu start menu. Then if I type one or two characters using the system keyboard / usr / share / X11 / xkb / symbols / bd, then the desired software shows up in the search results. But if I do the same thing with ibus-typing-booster or OpenBangla, then I have to finish typing the whole word, otherwise no software name appears in the search results. I have given two screenshots where I have tried to explain that issue.

![Screenshot from 2022-07-11 17-01-09](https://user-images.githubusercontent.com/60663471/178252658-2797319d-051e-4f0f-aafb-d6d612536da9.png)

![Screenshot from 2022-07-11 17-01-36](https://user-images.githubusercontent.com/60663471/178252718-d7f78565-c1cf-471c-854d-74c2e487f711.png)

**Comment by mike-fabian** [comment]
> Another thing I noticed. When I try to launch any software by searching in Ubuntu start menu. Then if I type one or two characters using the system keyboard / usr / share / X11 / xkb / symbols / bd, then the desired software shows up in the search results. But if I do the same thing with ibus-typing-booster or OpenBangla, then I have to finish typing the whole word, otherwise no software name appears in the search results. I have given two screenshots where I have tried to explain that issue.

This is not a problem of OpenBangla and ibus-typing-booster. It happens because these are input methods which have a preedit which the client (here the start menu) might not yet see until it is commited.

If you type a key with `/usr/share/X11/xkb/symbols/bd`, that key goes directly to the program typed into (here the start menu).

Input methods like OpenBangla and ibus-typing-booster have a preedit. The preedit might be indicated by an underline and/or colour depending on the toolkit and the settings of the input method.

But even if you turn the underline off, it is still a preedit. “preedit” means something like “unfinished edit”. The input methods are holding this back because something else might be coming which might change the result.

For example the `g` + `C` you mentioned, the `g` produces a "্" U+09CD BENGALI SIGN VIRAMA, then the `C` produces a "ৈ" U+09C8 BENGALI VOWEL SIGN AI and these need to be combined to ঐ U+0990 BENGALI LETTER AI. Therefore, the input method needs to hold back the "্" U+09CD BENGALI SIGN VIRAMA in the preedit and wait if something else is coming.

How much is held back in the preedit depends on the input method and what it tries to achieve.

A plain keyboard layout like `/usr/share/X11/xkb/symbols/bd` never does that, each key press immediately produces a final result, nothing is held back which makes more complicated stuff impossible with such a simple keyboard layout.

The reason why the start menu doesn’t show any results while the preedit is still there is that the start menu doesn’t see the preedit at all. As far as the start menu is concerned, there is still no input until the first “commit”. A “commit” means that some part of the input is finished and can be sent to the application (here the start menu). Only then  the start menu sees it.

Almost all programs behave like the start menu in this case, they do not see the contents of the preedit.

*But* I know at least one exception:

Google chrome sees apparently sees the contents of the preedit.

Look at this screenshot:

![Screenshot](https://user-images.githubusercontent.com/2330175/178265070-d241b13b-1208-4590-98c9-c5e73b77af92.png)

Here I am typing with ibus-typing-booster into the URL entry of Google Chrome. I typed `test` and it is still in preedit as shown clearly by the underline. ibus-typing-booster shows some candidate suggestions (The candidates labelled 1-6 on the right). But *below* the candidates shown by ibus-typing-booster, one sees candidates shown by the Google Chrome URL entry field! I.e. Google Chrome knows that there is a preedit and knows the contents of the preedit and can use already!

In this case, it is a bit of a conflict as I see both candidate lists at once, overlapping each other. But that is not a big problem, if I usually prefer the Google Chrome suggestions, I could configure ibus-typing-booster to pop up its candidate list only when I type Tab (In future I hope I can make such settings configurable for certain apps only, i.e. make it possible to require a Tab to shown the candidate list only when typing into Chrome but show the ibus-typing-booster candidate list immediately in other programs).

Google Chrome shows that it is possible that the application can know the contents of the preedit and react to it.

So in case of the start menu in Gnome (part of gnome-shell), I think gnome-shell could be improved to be able to see the preedit and do something with it.

I think this can only be fixed in gnome-shell, not in the input methods.

**Comment by Mahmudul-hasan613** [comment]
> > Also, when g + C is pressed, ঐ letter is supposed to appear. Which I can do in Open Bangla keyboard and bn-unijoy.mim layout. But that is not possible in the bn-national-jatiya.mim layout you created.
> 
> This is not the “automatic vowel forming”, this is something else, right?
> 
> And the /usr/share/X11/xkb/symbols/bd layout doesn’t support that either, right?
> 
> I think with `/usr/share/X11/xkb/symbols/bd` it is impossible to support that, but I should be able to improve bn-national-jatiya.mim to make that work.
> 
> I will try!

Yes, your are right. This is not the “automatic vowel forming”. And the /usr/share/X11/xkb/symbols/bd layout doesn’t support that either. But if I use bn-unijoy.mim as a plain keyboard, also this feature work.

**Comment by mike-fabian** [comment]
> But if I use bn-unijoy.mim as a plain keyboard, also this feature work.

bn-unijoy.mim is *not* a plain keyboard layout. It does sophisticated extra stuff

**Comment by Mahmudul-hasan613** [comment]
> bn-unijoy.mim is _not_ a plain keyboard layout. It does sophisticated extra stuff

Although "bn-unijoy.mim" is not a plain keyboard, it works much like a plain keyboard. If I type one or two characters in the Ubuntu Start menu through Unijoy, the search results come up.

**Comment by mike-fabian** [comment]
> > bn-unijoy.mim is _not_ a plain keyboard layout. It does sophisticated extra stuff
> 
> Although "bn-unijoy.mim" is not a plain keyboard, it works much like a plain keyboard. If I type one or two characters in the Ubuntu Start menu through Unijoy, the search results come up.

My guess is that you are not using it via ibus-typing-booster then. ibus-typing-booster delays the commit in order to be able to show possible suggestions for the rest of the word using a dictionary like `/usr/share/hunspell/bn_IN.dic` or using words you typed previously. So it cannot commit after typing only one or two characters, it tries to complete a word.

So I guess you are using `bn-unijoy.mim` not in ibus-typing-booster but via ibus-m17n.

ibus-m17n is very simple, it does not try to do any predictions. So ibus-m17n can commit already after the current letter is finished. It cannot commit "্" U+09CD BENGALI SIGN VIRAMA yet because it is  still waiting for the next letter, but if you then type "ৈ" U+09C8 BENGALI VOWEL SIGN AI, then the letter ঐ U+0990 BENGALI LETTER is finished. ibus-m17n then commits, there is nothing else to do. ibus-typing-booster and OpenBangla wait and keep the ঐ U+0990 BENGALI LETTER trying to do more stuff.

So if you don’t care for any extra stuff at all, you can also use bn-national-jatiya.mim via ibus-m17n instead of via ibus-typing-booster.

If you have my `bn-national-jatiya.mim` installed, you can add it in the Gnome settings like this:

The entry at the bottom of the list in that screenshot “(Bangla (national-jatiya (m17n))” is `bn-national-jatiya.mim` used via ibus-m17n:

![Screenshot](https://user-images.githubusercontent.com/2330175/178302773-93070117-02fc-451a-ae76-8ca87554056e.png)

**Comment by mike-fabian** [comment]
> > Also, when g + C is pressed, ঐ letter is supposed to appear. Which I can do in Open Bangla keyboard and bn-unijoy.mim layout. But that is not possible in the bn-national-jatiya.mim layout you created.
> 
> This is not the “automatic vowel forming”, this is something else, right?
> 
> And the /usr/share/X11/xkb/symbols/bd layout doesn’t support that either, right?
> 
> I think with `/usr/share/X11/xkb/symbols/bd` it is impossible to support that, but I should be able to improve bn-national-jatiya.mim to make that work.
> 
> I will try!

I have updated bn-national-jatiya.mim to support this:

https://github.com/mike-fabian/m17n-db-bn-national-jatiya/blob/main/bn-national-jatiya.mim

If you download this new file and put it in place of the one you already have, then log out of Gnome and log in again, you should have the new feature.

`g` acts as a “deadkey” now:

```
deadkey
  ("g" "্") ; U+09CD BENGALI SIGN VIRAMABENGALI SIGN VIRAMA
  )
```

and waits for the next key, if the next key is one of these:

```
 (independent
  ("h" (delete @-) "আ") ; + া U+09BE BENGALI VOWEL SIGN AA = আ U+0986 BENGALI LETTER AA
  ("d" (delete @-) "ই") ; + ি U+09BF BENGALI VOWEL SIGN I = ই U+0987 BENGALI LETTER I
  ("D" (delete @-) "ঈ") ; + ী U+09C0 BENGALI VOWEL SIGN II = ঈ U+0988 BENGALI LETTER II
  ("s" (delete @-) "উ") ; + ু U+09C1 BENGALI VOWEL SIGN U = উ U+0989 BENGALI LETTER U
  ("S" (delete @-) "ঊ") ; + ূ U+09C2 BENGALI VOWEL SIGN UU = ঊ U+098A BENGALI LETTER UU
  ("a" (delete @-) "ঋ") ; + ৃ U+09C3 BENGALI VOWEL SIGN VOCALIC R = ঋ U+098B BENGALI LETTER VOCALIC R
  ("c" (delete @-) "এ") ; + ে U+09C7 BENGALI VOWEL SIGN E = এ U+098F BENGALI LETTER E
  ("C" (delete @-) "ঐ") ; + ৈ U+09C8 BENGALI VOWEL SIGN AI = ঐ U+0990 BENGALI LETTER AI
  ("x" (delete @-) "ও") ; + ো U+09CB BENGALI VOWEL SIGN O = ও U+0993 BENGALI LETTER O
  ("X" (delete @-) "ঔ") ; + ৌ U+09CC BENGALI VOWEL SIGN AU = ঔ U+0994 BENGALI LETTER AU
  ("G" (delete @-) "॥") ; + । U+0964 DEVANAGARI DANDA = ॥ U+0965 DEVANAGARI DOUBLE DANDA
  ))

```

it deletes the "্" (“(delete @-)” deletes the previous character in the preedit) and inserts another character instead depending on which key was pressed.

If any other key not in this “independent” list is pressed, the "্" is kept as is and the other key is added after it. I checked that this is also the behaviour of OpenBangla.

**Comment by mike-fabian** [comment]
Is the following a bug in OpenBangla?:

If I press `a` in OpenBangla (Just `a`, *not* `AltGr+a`!), I get ঋ .

But https://github.com/OpenBangla/OpenBangla-Keyboard/blob/master/data/National_Jatiya.json  contains:

```
        "Key_a_AltGr": "ঋ",
        "Key_a_Normal": "ৃ",
```

i.e. without AltGr, the result should be "ৃ", not "ঋ".

Looking at the pictures shown by the OpenBangla setup tool for the National (Jatiya) layout, it also looks like that `a` should produce "ৃ".

**Comment by mike-fabian** [comment]
>  with bn-unijoy.mim, gnome settings do not need to be changed to use AltGr key when typing with them.

That is because `bn-unijoy.mim` uses the “normal” Alt key instead of AltGr. For example bn-unijoy.mim contains:

```
 ((A-u) "ঊ")	;; BENGALI LETTER UU
```

whereas bn-national-jatiya.mim contains:

```
 ((G-S) "ঊ") ; U+098A BENGALI LETTER UU

```

`A-u` means `Alt+u`, `G-S` means `AltGr+S`.

The standard “English (US)” keyboard layout has no `AltGr` key, the right and the left `Alt` key behave *exactly* the same in thne “English (US)” layout.

As bn-unijoy.mim uses `A-...` this means that you can use any keyboard layout which has an `Alt` key, you don’t need and `AltGr` key. But that can cause a lot of conflicts with keyboard shortcuts using Alt+something.

I think it is better to use `AltGr` for this and make sure that your keyboard layout has an `AltGr` key. Having `AltGr` on the right `Alt` key is more useful than to have just another duplicate of the left `Alt` key on the right side of the space bar. And it avoids a lot of conflicts with keyboard shortcuts starting with `Alt`.

**Comment by mike-fabian** [comment]
> > Is the following a bug in OpenBangla?:
> > If I press `a` in OpenBangla (Just `a`, _not_ `AltGr+a`!), I get ঋ .
> > But https://github.com/OpenBangla/OpenBangla-Keyboard/blob/master/data/National_Jatiya.json contains:
> > ```
> >         "Key_a_AltGr": "ঋ",
> >         "Key_a_Normal": "ৃ",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i.e. without AltGr, the result should be "ৃ", not "ঋ".
> > Looking at the pictures shown by the OpenBangla setup tool for the National (Jatiya) layout, it also looks like that `a` should produce "ৃ".
> 
> It's not a bug. It has done the work of Automatic Vowel Forming.

Ah, I see! If I switch that “Automatic Vowel Forming” off in OpenBangla, I get indeed "ৃ" when pressing `a`.

**Comment by mike-fabian** [comment]
Maybe I am beginning to understand what automatic vowel forming should do.

"ৃ" U+09C3 BENGALI VOWEL SIGN VOCALIC R

can probably be used to “apply” that vowel on something which came before. A consonant?

But if there is nothing before "ৃ" U+09C3 BENGALI VOWEL SIGN VOCALIC R it could apply to, either there is nothing at all as when I just type `a` where nothing else exists yet, or there is something but "ৃ" U+09C3 BENGALI VOWEL SIGN VOCALIC R cannot be applied to that something.

Then the  "ৃ" U+09C3 BENGALI VOWEL SIGN VOCALIC R transforms into a  ঋ U+098B BENGALI LETTER VOCALIC R

Do I understand that right?

Should any BENGALI VOWEL SIGN <something> transform into BENGALI LETTER <something> if there is nothing to the left  it could be applied to?

**Comment by Mahmudul-hasan613** [comment]
> Should any BENGALI VOWEL SIGN transform into BENGALI LETTER if there is nothing to the left it could be applied to?

Yes.

**Comment by Mahmudul-hasan613** [comment]
> can probably be used to “apply” that vowel on something which came before. A consonant?

Yes, consonants. To get a better idea about BENGALI VOWEL SIGN or VOWEL DIACRITICS, you can visit the [link](https://github.com/lifeparticle/Bengali-Alphabet/blob/master/vowel_diacritics.md).

**Comment by Mahmudul-hasan613** [comment]
> Then the "ৃ" U+09C3 BENGALI VOWEL SIGN VOCALIC R transforms into a ঋ U+098B BENGALI LETTER VOCALIC R

> Do I understand that right?

Yes, you got it right.

**Comment by mike-fabian** [comment]
Ah, nice. I try whether I can improve bn-national-jatiya.mim further to support that.

**Comment by Mahmudul-hasan613** [comment]
[Preedit disappears and appears elsewhere when the cursor is moved with a mouse.](https://github.com/OpenBangla/OpenBangla-Keyboard/issues/298) This bug is still ibus-typing-booster and OpenBangla Keyboard!


[Screencast from 13-7-22 09:32:47 পূর্বাহ্ণ +06.webm](https://user-images.githubusercontent.com/60663471/178646593-65963cb7-d6fe-4b88-9b16-619ee269f1cb.webm)

**Comment by mike-fabian** [comment]
It seems to me the behaviour for ibus-typing-booster and OpenBangla keyboard is different in your video. I think the difference is because OpenBangla does not support committing on focus change yet. So for OpenBangla, the preedit disappears and reappears elsewhere. For ibus-typing-booster, the preedit disappears in libreoffice, but does *not* reappear elsewhere. See my test video here:


https://user-images.githubusercontent.com/2330175/178718183-c56ed8bb-f0d0-402b-8902-e6aacfecb424.mp4

**Comment by mike-fabian** [comment]
That the preedit disappears for ibus-tying-booster as well, even though ibus-typing-booster supports commit on focus change, is probably a libreoffice bug.

**Comment by mike-fabian** [comment]
See this video using gedit for comparison:

In this video, when using OpenBangla, the preedit disappears and reappears at the new position, newly typed characters are added to the existing preedit at the new position.

But for ibus-typing-booster, clicking with the mouse to move the cursor to a different position commits the preedit at the old position (the underline disappears) and at the new position it does not  reappear, only the newly typed character is seen at the new position:

https://user-images.githubusercontent.com/2330175/178718652-e4c843fa-1a46-4645-8992-3991401127f2.mp4

**Comment by Mahmudul-hasan613** [comment]
> (At least not yet, OpenBangla 3.0.0 will support it apparently).

I am currently using 3.0.0 [develop](https://github.com/OpenBangla/OpenBangla-Keyboard/tree/develop) version of OpenBangla Keyboard. I mean, the new version is having this problem.

**Comment by mike-fabian** [comment]
Are you sure you are using 3.0.0? Does it display 3.0.0 when you open the “About OpenBangla Keyboard”?

When I use

```
bash -c "$(wget -q https://raw.githubusercontent.com/OpenBangla/OpenBangla-Keyboard/master/tools/install.sh -O -)"
```

or

```
bash -c "$(wget -q https://raw.githubusercontent.com/OpenBangla/OpenBangla-Keyboard/develop/tools/install.sh -O -)"
```

it installs the 2.0.0 version for me.

**Comment by Mahmudul-hasan613** [comment]
> Are you sure you are using 3.0.0? Does it display 3.0.0 when you open the “About OpenBangla Keyboard”?

I'm sure, I'm using version 3.0.0.
![Screenshot from 2022-07-16 07-34-56](https://user-images.githubusercontent.com/60663471/179329818-a87b91f2-e1d0-4ff3-ab09-d798ec88ed26.png)

**Comment by Mahmudul-hasan613** [comment]
> And when I try to build the version from the “develop” branch from source it fails for me as I describe here:
> 
> [OpenBangla/OpenBangla-Keyboard#307](https://github.com/OpenBangla/OpenBangla-Keyboard/discussions/307)

You can follow the method if you want.

> ```
> git clone --recursive https://github.com/OpenBangla/OpenBangla-Keyboard.git
> cd OpenBangla-Keyboard
> git checkout develop
> git submodule update
> mkdir build && cd build
> cmake .. -DENABLE_IBUS=ON
> make
> sudo make install
> ```

**Comment by mike-fabian** [comment]
OK, thank you, I was missing the `git submodule update`

**Comment by mike-fabian** [comment]
Here is a video using OpenBangla 3.0.0.

There is an improvement: 

when I type into gedit (above) an switch the focus to gnome-text-editor (below) the preedit is committed (i.e. does not disappear in gedit) and the input does not reappear. When I type into gnome-texteditor, and switch back to gedit, the preedit in gnome-text-editor is committed and does not disappear. 

But the following problem remains. 

When I typing into gedit and move the focus with the mouse to a different position inside gedit (i.e. inside the same window),  the preedit is correctly commited (with 2.0.0 it disappeared). So a slight improvement here as well.

*But* the following problems remain:

- the candidate list does not close
- the text reappears when typing at the new position, the newly typed text  is added to the old preedit

See this video:

https://user-images.githubusercontent.com/2330175/179356180-c1415f30-8c4c-4cfa-9f17-b092ab729de6.mp4

**Comment by mike-fabian** [comment]
New issue reported here:

https://github.com/OpenBangla/OpenBangla-Keyboard/issues/308

**Comment by Mahmudul-hasan613** [comment]
> New issue reported here:
> 
> [OpenBangla/OpenBangla-Keyboard#308](https://github.com/OpenBangla/OpenBangla-Keyboard/issues/308)

Thank you very much for working on the issue.

**Comment by mike-fabian** [comment]
I made a pull request for OpenBangla to fix the reappearing preedit when moving the cursor with the mouse to a different position in the same window:

https://github.com/OpenBangla/OpenBangla-Keyboard/issues/308#issuecomment-1215298089

**Comment by Mahmudul-hasan613** [comment]
Thank you very much for working on the issue.
When the pull request is done, will the problem be solved if I do a fresh install again...?

**Comment by mike-fabian** [comment]
> When the pull request is done, will the problem be solved if I do a fresh install again...?

Yes, when the pull request has been merged, the issue should be solved if you do a fresh install from the  develop branch.

**Comment by Mahmudul-hasan613** [comment]
> > When the pull request is done, will the problem be solved if I do a fresh install again...?
> 
> Yes, when the pull request has been merged, the issue should be solved if you do a fresh install from the develop branch.

Although the pull request has not been approved yet. But I have the same file as the develop branch; Added the code you suggested and did a fresh install. It is working very well. That problem no longer exists. Thank you very much.

**Comment by fujiwarat** [comment]
Seems not ibus issue.

**Comment by mike-fabian** [comment]
> Seems not ibus issue.

Yes!

**Comment by mike-fabian** [comment]
@Mahmudul-hasan613 
Can you please open a new issue at for the automatic vowel forming support?:
https://github.com/mike-fabian/m17n-db-bn-national-jatiya/issues
